src/array/cuda/rowwise_sampling.cu

```diff
@@ -232,27 +233,372 @@ __global__ void _CSRRowWiseSampleReplaceKernel(
+* @param out_ptr The offset to write each row to in the output COO.
+* @param out_rows The rows of the output COO (output).
+* @param out_cols The columns of the output COO (output).
+* @param out_idxs The data array of the output COO (output).
+*/
+template<typename IdType, typename FloatType, int TILE_SIZE>
+__global__ void _CSRRowWiseSampleKernel(
+    const uint64_t rand_seed,
+    const int64_t num_picks,
+    const int64_t num_rows,
+    const IdType * const in_rows,
+    const FloatType * const prob,
+    const IdType * const in_ptr,
+    const IdType * const in_index,
+    const IdType * const data,
+    const IdType * const out_ptr,
+    IdType * const out_rows,
+    IdType * const out_cols,
+    IdType * const out_idxs) {
+  // we assign one warp per row
+  assert(blockDim.x == BLOCK_SIZE);
+
+  int64_t out_row = blockIdx.x*TILE_SIZE+threadIdx.y;
+  const int64_t last_row = min(static_cast<int64_t>(blockIdx.x+1)*TILE_SIZE, num_rows);
+
+  curandStatePhilox4_32_10_t rng;
+  curand_init((rand_seed*gridDim.x+blockIdx.x)*blockDim.y+threadIdx.y, threadIdx.x, 0, &rng);
+
+  while (out_row < last_row) {
+    const int64_t row = in_rows[out_row];
+
+    const int64_t in_row_start = in_ptr[row];
+    const int64_t deg = in_ptr[row+1] - in_row_start;
+
+    const int64_t out_row_start = out_ptr[out_row];
+
+    if (deg <= num_picks) {
+      // just copy row
+      for (int idx = threadIdx.x; idx < deg; idx += BLOCK_SIZE) {
+        const IdType in_idx = in_row_start+idx;
+        out_rows[out_row_start+idx] = row;
+        out_cols[out_row_start+idx] = in_index[in_idx];
+        out_idxs[out_row_start+idx] = data ? data[in_idx] : in_idx;
+      }
+    } else {
+      // Algorithm A-Chao https://en.wikipedia.org/wiki/Reservoir_sampling#Algorithm_A-Chao
+      using BlockScanT = cub::BlockScan<FloatType, BLOCK_SIZE>;
+      // Allocate shared memory for BlockScan
+      __shared__ typename BlockScanT::TempStorage temp_storage;
+
+      FloatType weights_sum = static_cast<FloatType>(0.0);
+      // for the following two loops, we use a block scan window of size BLOCK_SIZE
+      // fill the reservoir array and compute the sum of their weights
+      for (int i = 0; i < (num_picks + BLOCK_SIZE - 1) / BLOCK_SIZE; i++) {
+        // Obtain input item for each thread
+        IdType idx = threadIdx.x + i * BLOCK_SIZE;
+        FloatType thread_prob;
+        if (idx < num_picks) {
+          _DoubleSlice<IdType, FloatType>(prob, data, idx, in_row_start, &thread_prob);
+          out_idxs[out_row_start+idx] = idx;
+        } else {
+          thread_prob = static_cast<FloatType>(0.0);
+        }
+        __syncthreads();
+
+        // Collectively compute the block-wide inclusive prefix sum
+        FloatType block_aggregate;
+        BlockScanT(temp_storage).InclusiveSum(thread_prob, thread_prob, block_aggregate);
+
+        weights_sum += block_aggregate;
+      }
+
+      // for items in [num_picks, deg), check if they can replace
```

Comment:
Typo in bracket

Reply:
I mean left-inclusive and right-exclusive here.